cpp/src/experimental/random_walks.cuh

```diff
@@ -0,0 +1,854 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Andrei Schaffer, aschaffer@nvidia.com
+//
+#pragma once
+
+#include <experimental/graph.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/graph_utils.cuh>
+
+#include <raft/device_atomics.cuh>
+#include <raft/handle.hpp>
+#include <raft/random/rng.cuh>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/copy.h>
+#include <thrust/find.h>
+#include <thrust/gather.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/tuple.h>
+
+#include <thrust/remove.h>
+#include <thrust/transform.h>
+
+#include <cassert>
+#include <ctime>
+#include <tuple>
+#include <type_traits>
+
+namespace cugraph {
+namespace experimental {
+
+namespace detail {
+
+template <typename T>
+using device_vec_t = rmm::device_uvector<T>;
```

Comment:
This is for 3 reasons: (1) testing-debugging: indirections like this to a a different type of container (say host; or a more debugging friendly wrapper around a device container) may help debugging an issue;
(2) If in the future an even better device container comes along (like `device_uvector` came as better alternative to `thrust::device_vector`) this helps updating the code by localizing the change in one place, only;
(4) hide `rmm` namespace dependency to make things less dependent on a very specialized implementation (that may change in the future).

Reply:
If we really need to introduce this indirection, this should be applied to the entire codebase (not just for random walks, if every algorithm brings their own conventions, it will make the entire codebase less readable and also make code maintenance much harder).

So, if you think this is really necessary for the codebase, let's discuss this with the team and we need to (at least plan to) apply this to the entire codebase.

Reply:
Let's take the example of switching from `thrust::device_vector` to `rmm::device_uvector`. It's not just a simple "replace-everywhere" in code, is it? 

Because `rmm::device_uvector` has a different interface: (i) it needs a cnstr. taking `stream`; (ii) it returns a raw pointer (`dv.data()`) differently than `thrust::device_vector` (`dv.data().get()`), possibly other issues. 

So, then _locally_ abstracting away the type of the device vector container helps avoiding potentially very many code changes. Switching the container is localized to a few changes in the preamble of the implementation / header file, as opposed to spread all over the place. 

And the abstraction is localized. It doesn't affect clients of this interface. And it also allows to switch back-n-forth between different containers for testing / debugging with minimal code changes.